salt/payload.py

```diff
@@ -132,9 +133,11 @@ def loads(self, msg, encoding=None):
                 # Due to this, if we don't need it, don't pass it at all so
                 # that under Python 2 we can still work with older versions
                 # of msgpack.
-                return msgpack.loads(msg, use_list=True, encoding=encoding)
+                ret = msgpack.loads(msg, use_list=True, encoding=encoding)
             else:
-                return msgpack.loads(msg, use_list=True)
+                ret = msgpack.loads(msg, use_list=True)
+            if six.PY3 and not raw:
+                ret = salt.transport.frame.decode_embedded_strs(ret)
```

Comment:
The original design had the caller of `loads` invoke `decode_embedded_strs`. If you add `decode_embedded_strs(ret)` here, then you should remove it from the callers (eg in 'transport/tcp.py' the callers that use `salt.crypt.Cryptical.loads` but not the ones that use the `msgpack.Unpacker`). Double decoding is inefficient. Or you could continue the tradition of having the caller responsible for this type of decoding. Your call.

Also, you don't want to run `decode_embedded_strs` when encoding is not None. The reason is that if encoding is not None, then it has been encoded with `use_bin_type=True`, in which case msgpack has already saved the distinction between `bytes` and `str`. `use_bin_type=True` is used in the IPC and events case since we don't care about breaking the wire protocol in those cases and it is much more efficient to have msgpack properly encode things in the first place.

eg the conditional would become:

`if six.PY3 and encoding is None and not raw:`